examples/C++/DDS/CustomListenerExample/CustomListenerPublisher.cpp

```diff
@@ -0,0 +1,170 @@
+
+    if (participant_ == nullptr)
+    {
+        return false;
+    }
+
+    //REGISTER THE TYPE
+    type_.register_type(participant_);
+
+    //CREATE THE PUBLISHER
+    publisher_ = participant_->create_publisher(PUBLISHER_QOS_DEFAULT, nullptr);
+
+    if (publisher_ == nullptr)
+    {
+        return false;
+    }
+
+    topic_ = participant_->create_topic("HelloWorldTopic", "HelloWorld", TOPIC_QOS_DEFAULT);
+
+    if (topic_ == nullptr)
+    {
+        return false;
+    }
+
+    // CREATE THE WRITER
+    StatusMask none_mask = StatusMask::none();
+    StatusMask all_mask = StatusMask::all();
+    dw_mask_ = use_dw ? all_mask : none_mask;
+    
+    writer_ = publisher_->create_datawriter(topic_, DATAWRITER_QOS_DEFAULT, &dw_listener_, dw_mask_);
+
+    if (writer_ == nullptr)
+    {
+        return false;
+    }
+    return true;
+}
+
+CustomListenerPublisher::~CustomListenerPublisher()
+{
+    if (writer_ != nullptr)
+    {
+        publisher_->delete_datawriter(writer_);
+    }
+    if (publisher_ != nullptr)
+    {
+        participant_->delete_publisher(publisher_);
+    }
+    if (topic_ != nullptr)
+    {
+        participant_->delete_topic(topic_);
+    }
+    DomainParticipantFactory::get_instance()->delete_participant(participant_);
+}
+
+void CustomListenerPublisher::runThread(
+        uint32_t samples,
+        uint32_t sleep)
+{
+    if (samples == 0)
+    {
+        while (!stop_)
+        {
+            if (publish(false))
+            {
+                std::cout << "Message: " << hello_.message() << " with index: " << hello_.index()
+                          << " SENT" << std::endl;
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(sleep));
+        }
+    }
+    else
+    {
+        for (uint32_t i = 0; i < samples; ++i)
+        {
+            if (!publish())
+            {
+                --i;
+            }
+            else
+            {
+                std::cout << "Message: " << hello_.message() << " with index: " << hello_.index()
+                          << " SENT" << std::endl;
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(sleep));
+        }
+    }
+}
+
+
+
+void CustomListenerPublisher::run(
+        uint32_t samples,
+        uint32_t sleep)
+{
+    stop_ = false;
+    std::thread thread(&CustomListenerPublisher::runThread, this, samples, sleep);
+    if (samples == 0)
+    {
+        std::cout << "Publisher running. Please press enter to stop the Publisher at any time." << std::endl;
+        std::cin.ignore();
+        stop_ = true;
+    }
+    else
+    {
+        std::cout << "Publisher running " << samples << " samples." << std::endl;
+    }
+    thread.join();
+}
+
+bool CustomListenerPublisher::publish(
+        bool waitForListener)
+{
+    if (dw_listener_.firstConnected_ || !waitForListener || dw_listener_.matched_ > 0 ||
+        p_listener_.firstConnected_  || p_listener_.matched_ > 0)
```

Comment:
I know this comes from the basic example, but `XX_listener_.matched_ > 0` is redundant with `XX_listener_.firstConnected_`, as the latter is set to true as soon as there is at least one match, and is never reverted back even if the number of matches comes to zero.
```suggestion
    if (!waitForListener || dw_listener_.firstConnected_ || p_listener_.firstConnected_)
```